README.md
1. USAGE
--------------

To run the Software Package from the command line using default parameters,
Run the run_predict_binding.py script as:
```shell
./run_predict_binding.py XXXX.pdb
```

2. INPUT
--------------

The three dimensional coordinates of the atomic positions of the protein
complex should be given in a PDB format.
Place the PDB files in the 'input' directory of the package.


3. OUTPUT
--------------

After a successful run of the program, the corresponding output files are
generated in the 'output' directory.
For each complex, two text files will be generated -
1. XXXX_scores.out - Contains the prediction and other metrics like z-score
2. XXXX_respairs.out - Contains the residue pairs comprising the interface

In case the Alanine Scanning flag is provided, an additional file will be generated,
3. XXXX_alascan.out - Contains the residues which upon mutation might destabilize the interaction


4. EXAMPLES
--------------

Examples for the input and output files are provided in the 'examples' directory


5. ADDITIONAL PARAMETERS
--------------

Optional flags that can be turned on -
USAGE: ./run_predict_binding.py XXXX.pdb -d 8.0 -t mm

1. Distance threshold for interaction definition: -d, --cutoff, choices = {4.0, 6.0, 8.0}
	NOTE: Default value = 4.0

2. Types of interacting atoms between residues: -t, --intertype, 
    choices = {'all', 'mm', 'ms', 'ss'}
	NOTE: Default value = ss
		  all - Interactions between all atoms
		  mm  - Interactions between main chain atoms
		  ms  - Interactions between main chain atoms and side chain atoms
		  ss  - Interactions between side chain atoms

3. To specify your own 20x20 matrix for residue-pair scores: -cp, --custom_pot, file
	NOTE: The custom potential should be in the format of a .csv file with the residue_pair in
		  first column and the corresponding score in the second column.
		  Example file provided in the 'examples' directory (custom_pot.csv)

4. To perform Alanine Scanning with the default potential: -alscn, --alascan, 
   choices = {'0', 1', '2', '3'}
	NOTE: Default value = '0'
		  0 - Do not perform Alanine Scanning
		  1 - Mutate all the residues on the interface to Alanine
		  2 - Mutate all the residues on the interface to every other residue
		  3 - Mutate a given list of residues to every other residue
		  (residues to be provided in a separate file)

	*NOTE: In case mode '3' is selected, a separate file containing the residues
		   to be mutated should be provided in the 'input' directory. An example
		   file is given in the 'examples' directory (4pza_alscan_input.txt).

5. In case of multimers, specify particular interfaces, -p1, -p2, --protein_1, --protein_2

6. To specify a name for the output files, -o, --outfile

This help can be accessed on the command line using:
./run_predict_binding.py -h or ./run_predict_binding.py --help


6. SOURCE CODE
----------------

All the scripts for the package can be found in the 'scripts' directory.


7. LICENSE
----------------

Please refer to COPYING.txt for the full license or COPYING_LESSER.txt for a condensed version.


CHANGELOG
README for PIZSA webserver edition v2.2

This is meant to serve as a changelog document noting all the additions made to PIZSA webserver source code, version 2.2 onwards

The source code can be found in 'scripts/', which is further subdivided into three directories to serve three major functions - 
'src_InterRes' - contains a subset of the cell_list algorithm scripts; computes the interface residues without running the prediction software
'src_main' - contains the cell_list algorithm as well as the prediction algorithm; computes all necessary parameters for PIZSA prediction
'src_mutanal' - contains code for alanine scanning predictions using PIZSA; can be used as a standalone program.

Additional functionalities are added as separated modules and older modules are deprecated when no longer needed.
Implementation of the additional modules is done in the 'predict_binding.py' script, beta-level changes will be implemented only in the 'predict_binding_interface' function 

Finding clashes - 
-- Modules added under 'src_main'
	-- find_clashes.py 
	-- percent_clash.py
	-- percent_clash_2.py 

-- find_clashes.py takes in the interacting residue pairs defined by a certain distance threshold and checks for any potential clashes.
	-- interacting atoms are classified as clashes if the distance between them is less than the sum of the van der Waal's radii of the constituent atoms
	-- 'Clash severity' is defined as the difference between the distance threshold ('cut_off') used for defining interactions and the interatomic distance

-- percent_clash.py takes in the clashes defined by the previous script and the cifa values computed elsewhere penalizing the clashing residues
	-- interacting residues classified as clashing are assigned a 'clash-severity' based on the fraction of interacting atoms involved in the clash
	-- clash penalty is computed through a sigmoidal function that takes 'clash-severity' as input
	-- DEPRECATED 

-- percent_clash_2.py takes in the clashes defined by the previous script and the cifa values defined elsewhere penalizing the clashing residues
	-- interacting atoms are classified as clashes according to the definition used in 'find_clashes.py'
	-- 'new_cifa' for clashing residues is computed by dividing the 'old_cifa' by the 'clash-severity'

Scaling scores for clashing residues - 
-- Modules added under 'src_main'
	-- score_complex.py 

-- score_complex_clash.py takes in a the maximum clashes between residue pairs on the interface and penalizes their scores 
	-- scale_dict output from percent_clash_2 returns the maximum atomic clash for each residue pair on the interface
	-- for any residue pair on the interface that has atomic clashes, a positive constant ( 1 in the current iteration) is added to the raw score
	-- the more numerous the clashes, the more unfavourable the raw score of the structure
	

RESULTS - 

After percent_clash_2.py - 
	Raw scores as well as Normalized Raw Scores are still more favourable for the decoys than the native structures

After score_complex_clash.py - 
	PDB_name, Stable, Unstable, Native
	t53, 0, 1355, Non-binder
	t50, 1451, 0, Binder (2 Bind, 1 Non-bind)
	t54, 0, 1387, Binder
	t41, 0, 1172, Non-binder
	t40, 0, 2128, Binder
	t47, 0, 1051, Non-binder
	t46, 0, 1555, Non-binder 
	t29, 0, 1972, Binder
	t38, 0, 874, Binder
	t39, 0, 1294, Binder
	t36, 0, 309, Non-binder 
	t37, 1129, 371, Binder (all)
	t35, 0, 499, Non-binder
	t32, 0, 599, Binder 
	t30, 0, 1201, Binder

	7 / 13 correct predictions for Native structures (t50 anq t37 are multimeric targets)


PIZSA webserver edition v2.4_beta

-- New modules : write_out_mm.py (Placeholder till I fix getting normalized z-scores for multimeric structures)

-- New functions/modifications:
	-- 'score_complex_clash.py' 
		-- This script now calculates normalized raw scores for the randomized decoys
	-- 'calc_zscore.py
		-- Added 'calc_zscore_norm' to calculate the z_scores based on normalized raw scores
	-- 'write_out.py'
		-- write_out function now takes the normalized output values and prints it in the main output file

scripts/src_main/addoptions.py
"""
Copyright (C) 2015 Abhilesh Dhawanjewar <abhilesh7@gmail.com>

This file is part of PIZSA.

PIZSA is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

PIZSA is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License along with PIZSA. If not, see <http://www.gnu.org/licenses/>.
"""

'''
This routine adds options to the main script for incorporating user inputs.
'''

# Provides the users flags to tweak the program
def addoptions():

	import argparse

	parser = argparse.ArgumentParser()
	parser.add_argument('input_pdb', help = 'Input PDB file for the protein complex')

	parser.add_argument('-d', '--cutoff', 
						type = float,
						choices = [4.0, 6.0, 8.0],
						default = 4.0,
						help = 'Specify the distance cut-off')

	parser.add_argument('-t', '--intertype',
						choices = ["all", "mm", "ms", "ss"],
						default = "ss", 
						help = 'Specify the interacting atom types')

	parser.add_argument('-o', '--outfile',
						help = 'Specify the output file')

	parser.add_argument('-cp', '--custom_pot',
						help = 'Specify the custom potential file in .csv format')

	parser.add_argument('-p1', '--protein_1',
						help = 'Specify the first protein of the interface')

	parser.add_argument('-p2', '--protein_2',
						help = 'Specify the second protein of the interface')

	parser.add_argument('-ma', '--mut_analysis',
						nargs = '*',
						default = '0',
						help = 'Specify whether Mutational Analysis is required')

	args = parser.parse_args()

	pdb_input = args.input_pdb
	cut_off = args.cutoff
	inter_type = args.intertype
	outfile = args.outfile
	protein_1 = args.protein_1
	protein_2 = args.protein_2
	custom_pot = args.custom_pot
	alscn = args.mut_analysis

	return pdb_input, cut_off, inter_type, protein_1, protein_2, outfile, custom_pot, alscn
